### ISSUE TITLE
js.UndefOr as alias for union broke Scala 2.12.0

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3024]
Because of a ~~type inference issue~~ scalac bug. 2.11.x and 2.12.1+ are fine.

Discovered by the nightly+weekly builds.

**Comment by sjrd** [comment]
It's actually a hard crash of scalac during the new `fields` phase of Scala 2.12:
```
[info] Compiling 108 Scala sources to /localhome/doeraene/projects/scalajs/library/target/scala-2.12/classes...
[error] error: scala.reflect.internal.Types$TypeError: type mismatch;
[error]  found   : scala.scalajs.js.|[Nothing,Unit]
[error]  required: scala.scalajs.js.|[Boolean,Unit]
[error] Note: Nothing <: Boolean, but trait | is invariant in type A.
[error] You may wish to investigate a wildcard type such as `_ <: Boolean`. (SLS 3.2.10)
[error]         at scala.tools.nsc.typechecker.Contexts$ThrowingReporter.handleError(Contexts.scala:1374)
[error]         at scala.tools.nsc.typechecker.Contexts$ContextReporter.issue(Contexts.scala:1226)
[error]         at scala.tools.nsc.typechecker.Contexts$Context.issue(Contexts.scala:575)
[error]         at scala.tools.nsc.typechecker.ContextErrors$ErrorUtils$.issueTypeError(ContextErrors.scala:106)
[error]         at scala.tools.nsc.typechecker.ContextErrors$ErrorUtils$.issueNormalTypeError(ContextErrors.scala:99)
[error]         at scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen$.AdaptTypeError(ContextErrors.scala:219)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adaptMismatchedSkolems$1(Typers.scala:1044)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.lastTry$1(Typers.scala:1055)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adaptExprNotFunMode$1(Typers.scala:1105)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.vanillaAdapt$1(Typers.scala:1151)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adapt(Typers.scala:1195)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5514)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5530)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5474)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5478)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.transformedOrTyped(Typers.scala:5709)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2270)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedMemberDef$1(Typers.scala:5412)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5463)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5500)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5530)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5474)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5478)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5555)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedPos(Typers.scala:5562)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.typedPos(Fields.scala:514)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.mkAccessor(Fields.scala:520)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:711)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformTermsAtExprOwner(Fields.scala:745)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.$anonfun$transformStats$3(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer$$Lambda$1398/870104058.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:513)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1408)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$1(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer$$Lambda$1091/577085522.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:513)
[error]         at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
[error]         at scala.reflect.internal.Trees.$anonfun$itransform$4(Trees.scala:1412)
[error]         at scala.reflect.internal.Trees$$Lambda$353/1384454980.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1411)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.$anonfun$transformStats$4(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer$$Lambda$1388/53120227.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:513)
[error]         at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1430)
[error]         at scala.reflect.internal.Trees$$Lambda$346/1768242710.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1430)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:42)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer$$Lambda$1090/1195815705.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:513)
[error]         at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
[error]         at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:402)
[error]         at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:395)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:366)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:366)
[error]         at scala.tools.nsc.Global$GlobalPhase$$Lambda$283/782689036.apply(Unknown Source)
[error]         at scala.collection.AbstractIterator.foreach(Iterator.scala:932)
[error]         at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:366)
[error]         at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1404)
[error]         at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1389)
[error]         at scala.tools.nsc.Global$Run.compileSources(Global.scala:1384)
[error]         at scala.tools.nsc.Global$Run.compile(Global.scala:1478)
[error]         at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
[error]         at scala.tools.nsc.Driver.process(Driver.scala:55)
[error]         at scala.tools.nsc.Driver.main(Driver.scala:68)
[error]         at scala.tools.nsc.Main.main(Main.scala)
[error] Exception in thread "main" scala.reflect.internal.Types$TypeError: type mismatch;
[error]  found   : scala.scalajs.js.|[Nothing,Unit]
[error]  required: scala.scalajs.js.|[Boolean,Unit]
[error] Note: Nothing <: Boolean, but trait | is invariant in type A.
[error] You may wish to investigate a wildcard type such as `_ <: Boolean`. (SLS 3.2.10)
[error]         at scala.tools.nsc.typechecker.Contexts$ThrowingReporter.handleError(Contexts.scala:1374)
[error]         at scala.tools.nsc.typechecker.Contexts$ContextReporter.issue(Contexts.scala:1226)
[error]         at scala.tools.nsc.typechecker.Contexts$Context.issue(Contexts.scala:575)
[error]         at scala.tools.nsc.typechecker.ContextErrors$ErrorUtils$.issueTypeError(ContextErrors.scala:106)
[error]         at scala.tools.nsc.typechecker.ContextErrors$ErrorUtils$.issueNormalTypeError(ContextErrors.scala:99)
[error]         at scala.tools.nsc.typechecker.ContextErrors$TyperContextErrors$TyperErrorGen$.AdaptTypeError(ContextErrors.scala:219)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adaptMismatchedSkolems$1(Typers.scala:1044)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.lastTry$1(Typers.scala:1055)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adaptExprNotFunMode$1(Typers.scala:1105)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.vanillaAdapt$1(Typers.scala:1151)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.adapt(Typers.scala:1195)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5514)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5530)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5474)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5478)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.transformedOrTyped(Typers.scala:5709)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2270)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedMemberDef$1(Typers.scala:5412)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5463)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.runTyper$1(Typers.scala:5500)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedInternal(Typers.scala:5530)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.body$2(Typers.scala:5474)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5478)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5555)
[error]         at scala.tools.nsc.typechecker.Typers$Typer.typedPos(Typers.scala:5562)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.typedPos(Fields.scala:514)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.mkAccessor(Fields.scala:520)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:711)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformTermsAtExprOwner(Fields.scala:745)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.$anonfun$transformStats$3(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer$$Lambda$1398/870104058.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:513)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1408)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$1(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer$$Lambda$1091/577085522.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:513)
[error]         at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
[error]         at scala.reflect.internal.Trees.$anonfun$itransform$4(Trees.scala:1412)
[error]         at scala.reflect.internal.Trees$$Lambda$353/1384454980.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1411)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:44)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.$anonfun$transformStats$4(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer$$Lambda$1388/53120227.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:758)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transformStats(Fields.scala:513)
[error]         at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1430)
[error]         at scala.reflect.internal.Trees$$Lambda$346/1768242710.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:25)
[error]         at scala.reflect.internal.Trees.itransform(Trees.scala:1430)
[error]         at scala.reflect.internal.Trees.itransform$(Trees.scala:1340)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
[error]         at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.super$transform(TypingTransformers.scala:40)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:42)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer$$Lambda$1090/1195815705.apply(Unknown Source)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:30)
[error]         at scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:25)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:737)
[error]         at scala.tools.nsc.transform.Fields$FieldsTransformer.transform(Fields.scala:513)
[error]         at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
[error]         at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:402)
[error]         at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:395)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:366)
[error]         at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:366)
[error]         at scala.tools.nsc.Global$GlobalPhase$$Lambda$283/782689036.apply(Unknown Source)
[error]         at scala.collection.AbstractIterator.foreach(Iterator.scala:932)
[error]         at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:366)
[error]         at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1404)
[error]         at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1389)
[error]         at scala.tools.nsc.Global$Run.compileSources(Global.scala:1384)
[error]         at scala.tools.nsc.Global$Run.compile(Global.scala:1478)
[error]         at scala.tools.nsc.MainClass.doCompile(Main.scala:24)
[error]         at scala.tools.nsc.Driver.process(Driver.scala:55)
[error]         at scala.tools.nsc.Driver.main(Driver.scala:68)
[error]         at scala.tools.nsc.Main.main(Main.scala)
java.lang.Exception: Nonzero exit code returned from runner: 1
        at ExternalCompile$$anonfun$4$$anonfun$7$$anonfun$ExternalCompile$$anonfun$$anonfun$$doCompile$1$1.apply(ExternalCompile.scala:80)
        at ExternalCompile$$anonfun$4$$anonfun$7$$anonfun$ExternalCompile$$anonfun$$anonfun$$doCompile$1$1.apply(ExternalCompile.scala:80)
        at scala.Option.foreach(Option.scala:236)
        at ExternalCompile$$anonfun$4$$anonfun$7.ExternalCompile$$anonfun$$anonfun$$doCompile$1(ExternalCompile.scala:80)
        at ExternalCompile$$anonfun$4$$anonfun$7.apply(ExternalCompile.scala:93)
        at ExternalCompile$$anonfun$4$$anonfun$7.apply(ExternalCompile.scala:44)
        at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
        at sbt.FileFunction$$anonfun$cached$1.apply(Tracked.scala:253)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:267)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3$$anonfun$apply$4.apply(Tracked.scala:263)
        at sbt.Difference.apply(Tracked.scala:224)
        at sbt.Difference.apply(Tracked.scala:206)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:263)
        at sbt.FileFunction$$anonfun$cached$2$$anonfun$apply$3.apply(Tracked.scala:262)
        at sbt.Difference.apply(Tracked.scala:224)
        at sbt.Difference.apply(Tracked.scala:200)
        at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:262)
        at sbt.FileFunction$$anonfun$cached$2.apply(Tracked.scala:260)
        at ExternalCompile$$anonfun$4.apply(ExternalCompile.scala:100)
        at ExternalCompile$$anonfun$4.apply(ExternalCompile.scala:18)
        at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
        at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
        at sbt.std.Transform$$anon$4.work(System.scala:63)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
        at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
        at sbt.Execute.work(Execute.scala:237)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
        at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
        at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
[error] (library/compile:compile) java.lang.Exception: Nonzero exit code returned from runner: 1
```
Definitely looks like a severe scalac bug, that was fixed in 2.12.1. Among [the PRs mentioning "fields" that made it into 2.12.1](https://github.com/scala/scala/pulls?utf8=%E2%9C%93&q=is%3Apr%20fields%20is%3Aclosed%20milestone%3A2.12.1), the only reasonable candidate would be https://github.com/scala/scala/pull/5554.

Since I don't even have a file name or anything, I'm not sure we can do anything about it.

@gzm0 Would it be reasonable to say we simply don't support 2.12.0 in 1.x, due to a regression on their side?

**Comment by gzm0** [comment]
How widespread is this issue? Do only certain union types not work or none at all?

I'm in principal fine with dropping support for 2.12.0 but we should:
- Ask the community whether there are reasons not to do that (e.g. a difficult 2.12.0 -> 2.12.1 migration path). Just to be double safe.
- Investigate whether it is feasible to just not support the things that fail (and deactivate the tests), rather than dropping support entirely.

**Comment by sjrd** [comment]
> How widespread is this issue? Do only certain union types not work or none at all?

A bit of `git grep`'ed-based debugging suggests that it is systematic for fields with a union type in *traits*. The crash above originates from the very definition of `js.PropertyDescriptor`. If I comment out the fields there, I receive another similar crash, whose origin I have not tracked down.

I have a minimization that is Scala.js-independent here: https://scastie.scala-lang.org/sjrd/4248wt0yQyC9KjxwPfwOlg/1
```scala
import scala.annotation.unchecked.uncheckedVariance

trait Union[A, B]

object Undef {
  type UndefOr[+A] = Union[A @uncheckedVariance, Unit]

  val undefined: UndefOr[Nothing] = null
}

trait Foobar {
  var field: Undef.UndefOr[String] = Undef.undefined
}
```
That's basically the shape of optional fields in non-native JS traits. They're systematically broken in 2.12.0.

The reason appears to be that the `typer` is happy with the subtyping relationship between `undefined` of type `UndefOr[Nothing]` and `UndefOr[String]` because `UndefOr` is covariant. But the fields phase dealiases the types into `Union[Nothing, Unit]` and `Union[String, Unit]`, but since `Union` is nonvariant, it produces nonsensical code. It should introduce the necessary casts to fix it up, which is what 2.12.1 apparently does but 2.12.0 does not.

**Comment by sjrd** [comment]
> Investigate whether it is feasible to just not support the things that fail (and deactivate the tests), rather than dropping support entirely.

At least this has a clear No answer. The `library` does not compile at all.

**Comment by gzm0** [comment]
Ok. So if we have to chose between 2.12.0 and `UndefOr[T] = T | Unit` I also chose the latter. And indeed by the time we have released SJS 1.0, 2.12 should be a couple of patch versions forward, so we should be good.
